### PR TITLE
Manage reservation

### PR DESCRIPTION
**Pull request by JustAnOtherCoder13**
review model to match the new logic. Get a reference of currentUser in rerstaurant view model, certainly not the good way to do, but for now i Just want to know about daoImpl, repo/interactors and view model/view if the work done is in the good way.

### CODE REVIEW

File: app/src/main/java/com/picone/go4lunch/presentation/ui/AuthenticationFragment.java [hunk 1]
```diff
@@ -55,9 +55,9 @@ public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup c
     public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
         super.onViewCreated(view, savedInstanceState);
         initAuthenticationAborted();
-        mUserViewModel.getCurrentUser().observe(getViewLifecycleOwner(), currentUser -> {
+        mUserViewModel.getCurrentUser.observe(getViewLifecycleOwner(), currentUser -> {
             if (isNewUser) mUserViewModel.addUser(currentUser);
-            mUserViewModel.getAddUserState().observe(getViewLifecycleOwner(), addUserState -> {
```
Comment (fodeguirassy) [inline app/src/main/java/com/picone/go4lunch/presentation/ui/AuthenticationFragment.java:60]: Nested livedata subscription.

File: app/src/main/java/com/picone/go4lunch/presentation/ui/fragment/RestaurantListFragment.java [hunk 2]
```diff
@@ -41,9 +45,12 @@ public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup c
         return mBinding.getRoot();
     }
 
+    @SuppressWarnings("ConstantConditions")
+    //mAuth.getCurrentUser().getEmail() can't be null cause needed to enter the app
     @Override
     public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
         super.onViewCreated(view, savedInstanceState);
```
Comment (fodeguirassy) [inline app/src/main/java/com/picone/go4lunch/presentation/ui/fragment/RestaurantListFragment.java:45]: Why are you querying the User as it's already known to LoginViewModel and UserViewModel

File: app/src/main/java/com/picone/go4lunch/presentation/viewModels/RestaurantViewModel.java [hunk 3]
```diff
@@ -1,32 +1,211 @@
...
+                    }
+
+                    @Override
+                    public void onError(Throwable e) {
+
+                    }
+                });
+    }
+
+    private void setDate() {
```
Comment (fodeguirassy) [inline app/src/main/java/com/picone/go4lunch/presentation/viewModels/RestaurantViewModel.java:143]: today starts as a Date the it is initialized again from a SimpleDateFormat Doesn't make sense

File: core/src/main/java/com/picone/core/data/repository/RestaurantDaoImpl.java [hunk 4]
```diff
@@ -25,4 +48,60 @@ public RestaurantDaoImpl(FirebaseDatabase database){
...
+                                    .addOnFailureListener(
+                                            e -> {
+                                                if (!emitter.isDisposed())
+                                                    emitter.onError(e);
+                                            });
+                        }
+                    }
+
+                    @Override
+                    public void onCancelled(@NonNull DatabaseError error) {
```
Comment (fodeguirassy) [inline core/src/main/java/com/picone/core/data/repository/RestaurantDaoImpl.java:74]: emitter.onError ?

File: app/src/main/java/com/picone/go4lunch/presentation/viewModels/RestaurantViewModel.java [hunk 5]
```diff
@@ -1,32 +1,211 @@
...
+                        updateUserForRestaurant(restaurant, restaurantsForName.get(0).getKey());
+                    }
+                });
+    }
+
+    private void addRestaurantWhenNotPersisted(Restaurant restaurant) {
+        restaurant.setRestaurantDailySchedule(new RestaurantDailySchedule(DATE, new ArrayList<>()));
+        addRestaurantInteractor.addRestaurant(restaurant)
+                .subscribeOn(Schedulers.io())
+                .observeOn(AndroidSchedulers.mainThread())
```
Comment (fodeguirassy) [inline app/src/main/java/com/picone/go4lunch/presentation/viewModels/RestaurantViewModel.java]: Why subscribe isn't enough here as you've already managed to do with other observables ?

**Review by fodeguirassy — CHANGES_REQUESTED** [review]

File: app/src/main/java/com/picone/go4lunch/presentation/ui/fragment/RestaurantListFragment.java (hunk 2 above)
Comment (JustAnOtherCoder13) [reply app/src/main/java/com/picone/go4lunch/presentation/ui/fragment/RestaurantListFragment.java:45]: Login View model doesn't know the current user, it's just an authentication state view model, and in the UserViewModel i set current user with the value of a new user to add it in db, but i never request the user in db. If i don't do this request, i never get access to user daily schedule. Perhaps there is a way to change this and do more simple.

File: app/src/main/java/com/picone/go4lunch/presentation/viewModels/RestaurantViewModel.java (hunk 3 above)
Comment (JustAnOtherCoder13) [reply app/src/main/java/com/picone/go4lunch/presentation/viewModels/RestaurantViewModel.java:143]: ok resolved

File: core/src/main/java/com/picone/core/data/repository/RestaurantDaoImpl.java (hunk 4 above)
Comment (JustAnOtherCoder13) [reply core/src/main/java/com/picone/core/data/repository/RestaurantDaoImpl.java:74]: I have made the same way Rx2Firebase done, i have show how they write their setValue method and adapt it to my usage.
So i think it is to throw an error on writing dailyShedule.

File: app/src/main/java/com/picone/go4lunch/presentation/viewModels/RestaurantViewModel.java (hunk 5 above)
Comment (JustAnOtherCoder13) [reply app/src/main/java/com/picone/go4lunch/presentation/viewModels/RestaurantViewModel.java]: Ok i didn't know that, if i understand, if my subscribeOn() and observeOn have been set in "parent" method that call other method, i don't have to set them again in "child" method?

File: app/src/main/java/com/picone/go4lunch/presentation/ui/AuthenticationFragment.java (hunk 1 above)
Comment (JustAnOtherCoder13) [reply app/src/main/java/com/picone/go4lunch/presentation/ui/AuthenticationFragment.java:60]: the things i have seen on the net call of mediator live data or Transformations.switchMap method in view model. Is this what you're thinking about?

Comment (JustAnOtherCoder13) [reply app/src/main/java/com/picone/go4lunch/presentation/ui/AuthenticationFragment.java:60]: Done

File: core/src/main/java/com/picone/core/data/repository/RestaurantDaoImpl.java (hunk 4 above)
Comment (JustAnOtherCoder13) [reply core/src/main/java/com/picone/core/data/repository/RestaurantDaoImpl.java:74]: Can't refactor this method without refactoring addUser.

. If i push firebase create a new user.
. If i don't push, firebase remove all users and write new user in root.

To resolve this i have to not push user when added, but write under a node like "name" or "uid".

Do i have to make changes?

File: app/src/main/java/com/picone/go4lunch/presentation/viewModels/RestaurantViewModel.java (hunk 5 above)
Comment (JustAnOtherCoder13) [reply app/src/main/java/com/picone/go4lunch/presentation/viewModels/RestaurantViewModel.java]: Done pass by flatMap, use andThen to combine observable on completable.

Find a more concise way to throw exeptions, just have to had a lambda throwable in subscribe.

File: core/src/main/java/com/picone/core/data/repository/RestaurantDaoImpl.java (hunk 4 above)
Comment (fodeguirassy) [reply core/src/main/java/com/picone/core/data/repository/RestaurantDaoImpl.java:74]: `databaseref.child('users').child(currentUser.getKey().setValue(currentUser))`

File: app/src/main/java/com/picone/go4lunch/presentation/viewModels/RestaurantViewModel.java [hunk 6]
```diff
@@ -1,32 +1,147 @@
...
+        selectedRestaurantMutableLiveData.setValue(selectedRestaurant);
+        getRestaurantForNameInteractor.getRestaurantForName(selectedRestaurant.getName())
+                .flatMap((Function<List<Restaurant>, ObservableSource<List<User>>>)
+                        restaurantsForName -> {
+                            String restaurantKey;
+                            if (!restaurantsForName.isEmpty()) {
+                                restaurantKey = restaurantsForName.get(0).getKey();
+                                selectedRestaurantKeyMutableLiveData.setValue(restaurantKey);
+                            } else {
+                                restaurantKey = "unknown";
```
Comment (fodeguirassy) [inline app/src/main/java/com/picone/go4lunch/presentation/viewModels/RestaurantViewModel.java:93]: Chain this below database operation

**Review by fodeguirassy — CHANGES_REQUESTED** [review]

File: core/src/main/java/com/picone/core/data/repository/RestaurantDaoImpl.java (hunk 4 above)
Comment (JustAnOtherCoder13) [reply core/src/main/java/com/picone/core/data/repository/RestaurantDaoImpl.java:74]: So to do that way i have to add a parameter "key" in User model in order to get the key.

In firebase the getKey method could only be access with the push method, so i could only do this when i push user.

**Pull request merged**